core/src/main/java/io/grpc/ServerInterceptors.java

```diff
@@ -80,6 +84,50 @@ public static ServerServiceDefinition intercept(ServerServiceDefinition serviceD
     return serviceDefBuilder.build();
   }
 
+  /**
+   * Create a new {@code ServerServiceDefinition} whose {@link MethodDescriptor} serializes to
+   * and from byte[] for all methods.  The {@code ServerCallHandler} created will automatically
+   * convert back to the original types for request and response before calling the existing
+   * {@code ServerCallHandler}.  Calling this method combined with the intercept methods will
+   * allow the developer to choose whether to intercept messages of byte[], or the modeled types
+   * of their application.
+   *
+   * @param serviceDef the service definition to convert messages to byte[]
+   * @return a wrapped version of {@code serviceDef} with the byte[] conversion applied.
+   */
+  public static ServerServiceDefinition useByteArrayMessages(
+      final ServerServiceDefinition serviceDef) {
+    final MethodDescriptor.Marshaller<byte[]> marshaller =
```

Comment:
The more I think about it, the more I think `InputStream` is the way to go, since it has the most optimization opportunities and this is on a speed-critical path. I fully agree it is more annoying to write the interceptor then, but it seems to provide the greatest flexibility for implementations and performance.

It's true that `InputStream` is not guaranteed to support mark, but if `markSupported() == false` then you could wrap it in a `BufferedInputStream` which does support mark. Using `BufferedInputStream` would be similar memory usage to using `byte[]`.

`byte[]` is a quite limited API. You can't for instance just remove one byte from the beginning or end; you have to make a copy that is different by one byte. It also isn't generally clear whether you can modify the `byte[]`, so often you need to make possibly-unnecessary copies. It also must be a single contiguous region of memory, which is quite a problem if a message may be 100 or even 10 MB. This is especially true since you tend to have to read into `byte[]`s to get all the bytes and then copy into a new `byte[]` that is the perfect size.

I'll be the first to admit that `InputStream` has its own deficiencies, but it has one key advantage: we can extend it to add additional methods, like done with `Drainable`. That provides a lot of optimization opportunities that can be used in `MethodDescriptor.Marshaller`s as well.